### PR TITLE
Add "Clear Puzzle" button to reset workspace to start_blocks

### PR DESCRIPTION
**Pull request by joshlory**
This change adds a "Clear Puzzle" button above the trashcan that resets the workspace to a clean slate.  In order to make room for the button, the main puzzle area header now says "Workspace:" instead of "Assemble your blocks here:".

### CODE REVIEW

**Comment by breville** [comment]
very cool

**Comment by cpirich** [comment]
We should consider whether this term (puzzle) makes sense for freePlay or project levels.

Also, without Undo, this could be a very destructive action when building complex creations. Is there a confirmation dialog?

Lastly, we may want to conditionally hide this altogether while in free-text (ace editor) JavaScript mode. It seems like it is not that hard to delete text in the editor. Sarah should have an opinion.

Thanks

**Comment by joshlory** [comment]
Great points.  The "Clear Puzzle" button shouldn't show up on droplet levels (if it does that's a bug).

Context here is that we're adding autosave to all levels, and the user needs a way to get back to the original start blocks vs. what was saved from a previous attempt.  @sfilman, I wonder if the original plan might be less confusing for the users (replacing the button in the instructions with "Load previous work" and "Clear Puzzle").  Otherwise when they load a level they haven't worked on for a while it may not be clear that the blocks they added previously aren't start blocks.

Totally agree on undo/redo.  When we last looked this wasn't a trivial feature to add.  Sarah can make the call for whether this needs to be a prerequisite for "Clear Puzzle" or whether adding a confirmation dialog is enough.

**Comment by sfilman** [comment]
@joshlory and I talked offline -- we are going to:
- suppress the button from the droplet levels.
- add a confirmation dialog to the clear puzzle action. Undo/redo stack is out of scope for this milestone, so the confirmation dialog will put an extra barrier up for this destructive action.
- wait on any additional UI affordance to let the user know that we've loaded their previous solution, as it's unclear that it's an important problem to solve for yet.

**Comment by joshlory** [comment]
screen shot 2015-02-13 at 9 02 51 pm

**Comment by sfilman** [comment]
Looks good! Nit: Can we do lower case on puzzle in the button?

**Comment by sfilman** [comment]
Mm seems like we have different standards for this in various places. Let's
leave it alone for now.

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -783,6 +794,10 @@ StudioApp.prototype.resizeHeaders = function (fullWorkspaceWidth) {
   var toolboxWidth = 0;
   var showCodeWidth = 0;
 
+  var clearPuzzleHeader = document.getElementById('clear-puzzle-header');
+  var clearPuzzleWidth = clearPuzzleHeader
+      ? clearPuzzleHeader.getBoundingClientRect().width : 0;
```
Comment (Bjvanminnen) [inline apps/src/StudioApp.js]: Nit, ? should be on the previous line. Style guide even backs me up :)
http://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml#Code_formatting (see last section on binary and ternary operators)

File: apps/src/StudioApp.js [hunk 2]
```diff
@@ -1217,6 +1232,19 @@ StudioApp.prototype.setCheckForEmptyBlocks = function (checkBlocks) {
 };
 
 /**
+ * Add the starting block(s).
+ */
+StudioApp.prototype.setStartBlocks_ = function (config) {
+  var startBlocks = config.level.startBlocks || '';
+  if (config.forceInsertTopBlock) {
+    startBlocks =
+        blockUtils.forceInsertTopBlock(startBlocks, config.forceInsertTopBlock);
```
Comment (Bjvanminnen) [inline apps/src/StudioApp.js]: I personally prefer something like
```
startBlocks = blockUtils.forceInsertTopBlock(startBlocks,
    config.forceInsertTopBlock)
```
but it does look like style guide says how you have it is okay too.

File: apps/src/feedback.js [hunk 3]
```diff
@@ -745,6 +745,48 @@ FeedbackUtils.prototype.showGeneratedCode = function(Dialog) {
 /**
  *
  */
+FeedbackUtils.prototype.showClearPuzzleConfirmation = function(Dialog, callback) {
```
Comment (Bjvanminnen) [inline apps/src/feedback.js]: Maybe a comment explaining what callback is/when it gets called.

**Comment by Bjvanminnen** [comment]
Couple very minor nits, but LGTM

File: apps/src/StudioApp.js (hunk 1 above)
Comment (joshlory) [inline apps/src/StudioApp.js]: I thought we added an exception :)

"Binary and Ternary line continuations should have the operator at the beginning of the line: (note that this is different from Google’s JavaScript style guide)"

https://docs.google.com/a/code.org/document/d/1N18MR7rY3gASdSPGOC0sKISNXxyoGGS59i3JWbteWOY/edit#heading=h.yzziwdeuup14

I'm open to either style.

File: apps/src/StudioApp.js (hunk 2 above)
Comment (joshlory) [inline apps/src/StudioApp.js]: Good catch, I like that better too.

**Pull request merged**